Wrappers/Python/cil/optimisation/functions/SAGFunction.py

```diff
@@ -0,0 +1,220 @@
+                1., self._full_gradient_at_iterate, 1., out=out)  
+
+        return out 
+    
+    def warm_start_approximate_gradients(self, initial):
+        """A function to warm start SAG or SAGA algorithms by initialising all the gradients at an initial point. Equivalently setting :math:`g_i^0=\nabla f_i(x_0)` for initial point :math:`x_0`. 
+        
+        Parameters
+        ----------
+        initial: DataContainer,
+            The initial point to warmstart the calculation
+            
+        Note
+        ----
+        When using SAG or SAGA with a deterministic algorithm, you should warm start the SAG-SAGA Function with the same initial point that you initialise the algorithm
+        
+        """
+        self._list_stored_gradients = [
+            fi.gradient(initial) for fi in self.functions]
+        self._full_gradient_at_iterate = np.sum(self._list_stored_gradients)
+        self._update_data_passes_indices(list(range(self.num_functions)))
+        self._sampled_grad = initial.copy()
+        self._stochastic_grad_difference = initial.copy()
+
+    @property
+    def data_passes_indices(self): 
+        """ The property :code:`data_passes_indices` is a list of lists holding the indices of the functions that are processed in each call of `gradient`. This list is updated each time `gradient` is called by appending a list of the indices of the functions used to calculate the gradient.  
+        This is overwritten from the base class to first check to see if the approximate gradient was warm started and, if it was, ensure that the first element of `data_passes_indices` contains each index used to warm start and the index used in the first call to `gradient`. Thus the length of `data_passes_indices` is always equal to the number of calls to `gradient`. 
+        """
+        ret = self._data_passes_indices[:]  
+        if len(ret[0]) == self.num_functions:  
+            a = ret.pop(1)  
+            ret[0] += a  
+        return ret
+    
+class SAGAFunction(SAGFunction):
+
+    r"""
+    SAGA (SAG-Ameliore) is an accelerated version of the stochastic average gradient (SAG) function which takes a index :math:`i_k` and calculates the approximate gradient of :math:`\sum_{i=1}^{n-1}f_i` at iteration :math:`x_k` as
+    
+    .. math ::
+                 n\left(g_{i_k}^{k}-g_{i_k}^{k-1}\right)+\sum_{i=1}^{n-1} g_i^{k-1} \qquad \text{where} \qquad g_i^k= \begin{cases}
+                                                                            \nabla f_i(x_k), \text{ if } i=i_k\\
+                                                                            g_i^{k-1},\text{ otherwise}
+                                                                            \end{cases}
+                                                                        
+         
+    SAGA improves on the theory behind SAG and SVRG, with better theoretical convergence rates. Compared to SAG it is an unbiased estimator. 
```

Comment:
SVRG?

Reply:
It's what they claim in the paper abstract: https://www.di.ens.fr/~fbach/Defazio_NIPS2014.pdf